crates/sui-indexer/migrations_v2/2023-08-19-044035_tx_indices/up.sql

```diff
@@ -0,0 +1,19 @@
+CREATE TABLE tx_indices (
+    tx_sequence_number          BIGINT       PRIMARY KEY,
+    transaction_digest          bytea        NOT NULL,
+    input_objects               bytea[]      NOT NULL,
+    changed_objects             bytea[]      NOT NULL,
+    senders                     bytea[]      NOT NULL,
+    recipients                  bytea[]      NOT NULL,
+    packages                    bytea[]      NOT NULL,
+    package_modules             text[]       NOT NULL,
+    package_module_functions    text[]       NOT NULL
+);
+
+CREATE INDEX tx_indices_input_objects ON tx_indices USING GIN(input_objects);
+CREATE INDEX tx_indices_chagned_objects ON tx_indices USING GIN(changed_objects);
+CREATE INDEX tx_indices_senders ON tx_indices USING GIN(senders);
+CREATE INDEX tx_indices_recipients ON tx_indices USING GIN(recipients);
+CREATE INDEX tx_indices_package ON tx_indices USING GIN(packages);
+CREATE INDEX tx_indices_package_module ON tx_indices USING GIN(package_modules);
+CREATE INDEX tx_indices_package_module_function ON tx_indices USING GIN(package_module_functions);
```

Comment:
By being careful about how the package module and function are encoded, can we collapse this down to one index? I.e. if we have a sentinel "start" and "end" token (like `^`, and `$`), will we be able to use a GIN index to look for `^package::`, `^package::module::` and `^package::module::function$`?

Reply:
Based on my research, I don't think it's possible to do prefix searching w/ GIN index in an array unfortunately. As you can guess, `package` is Object ID in bytes, `package_module` is like "0x2::foo", and `package_module_function` is "0x2::foo::bar". Fortunately the index size is not huge despite the duplicated data

Reply:
Yeah, I think any scheme I could think of is not really different from storing three separate posting lists!